### PR TITLE
Add `icon_modulate` set/get functionality to PopupMenu

### PR DESCRIPTION
**Pull request by snibk**
Copies the ``get_item_icon_modulate`` and ``set_item_icon_modulate`` functionality already present in the ``ItemList`` class over to ``PopupMenu``. I need this to be able to give different colors to items that have the same icon in a dropdown, so I'm guessing others may find the functionality useful as well?

image

### CODE REVIEW

**Review by bruvzg — COMMENTED** [review]
New methods should be exposed to the GDScript (add the following code to the `PopupMenu::_bind_methods`) and documented (see [Contributing to the class reference](https://docs.godotengine.org/en/latest/community/contributing/updating_the_class_reference.html#updating-the-documentation-template)).

```
ClassDB::bind_method(D_METHOD("set_item_icon_modulate", "index", "modulate"), &PopupMenu::set_item_icon_modulate);
ClassDB::bind_method(D_METHOD("get_item_icon_modulate", "index"), &PopupMenu::get_item_icon_modulate);
```

File: scene/gui/popup_menu.cpp [hunk 1]
```diff
@@ -1172,6 +1175,22 @@ void PopupMenu::set_item_icon(int p_idx, const Ref<Texture2D> &p_icon) {
...
+	}
+	ERR_FAIL_INDEX(p_idx, items.size());
+
+	if (items[p_idx].icon_modulate == p_modulate) {
+		return;
+	}
+
+	items.write[p_idx].icon_modulate = p_modulate;
+	control->queue_redraw();
+	child_controls_changed();
```
[the remark this answers is not in the document]
Comment (the-sink) [reply scene/gui/popup_menu.cpp]: Thanks, should be fixed in c304dd7!

**Comment by Chaosus** [comment]
@the-sink You need to squash the commits (see https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#modifying-a-pull-request)

**Comment by the-sink** [comment]
@Chaosus Rebased, and... reading through all the contributing docs is probably in order! Appreciate it.

File: scene/gui/popup_menu.cpp [hunk 2]
```diff
@@ -612,6 +612,9 @@ void PopupMenu::_draw_items() {
 
 		Color icon_color(1, 1, 1, items[i].disabled && !items[i].separator ? 0.5 : 1);
 
+		// Multiply icon color by the custom modulate color
```
Comment (KoBeWi) [inline scene/gui/popup_menu.cpp:615]:
```suggestion
		// Multiply icon color by the custom modulate color.
```
Although I think this comment isn't necessary.

**Comment by KoBeWi** [comment]
The PR looks ok functionality-wise, but it's not linked to any proposal (although I get how parity with ItemList somewhat justifies it).

I think it's missing being exposed to the inspector; you can only set modulate from code. Check how other properties are exposed and you could do the same.

**Comment by YuriSizov** [comment]
> The PR looks ok functionality-wise, but it's not linked to any proposal

It's for https://github.com/godotengine/godot-proposals/issues/768, to at least unlock that ability so the proposal can be implemented as a plugin.

File: scene/gui/popup_menu.cpp (hunk 2 above)
Comment (the-sink) [reply scene/gui/popup_menu.cpp:615]: Thanks, I agree the comment is unnecessary - will go ahead and remove it.

**Comment by the-sink** [comment]
> I think it's missing being exposed to the inspector; you can only set modulate from code. Check how other properties are exposed and you could do the same.

Referring to the `icon_modulate` for individual PopupMenu items? I agree that would be good to expose to the inspector, although as far as I can tell ItemList doesn't expose that option either. Do you think it would be better to expose the parameter for both classes in a separate PR or just do it here?

**Comment by KoBeWi** [comment]
> Do you think it would be better to expose the parameter for both classes in a separate PR or just do it here?

I remember ItemList has many properties that were not exposed to not bloat the list, so maybe it's better to not do it at all. I'd wait until there is some actual demand.

**Review by KoBeWi — APPROVED** [review]
Looks fine, aside from the comment I commented about.

**Comment by the-sink** [comment]
Should be good now!

File: scene/gui/popup_menu.cpp [hunk 3]
```diff
@@ -1249,6 +1251,15 @@ void PopupMenu::set_item_icon_max_width(int p_idx, int p_width) {
 	_menu_changed();
 }
 
+void PopupMenu::set_item_icon_modulate(int p_idx, const Color &p_modulate) {
+	if (items[p_idx].icon_modulate == p_modulate) {
```
Comment (akien-mga) [inline scene/gui/popup_menu.cpp:1255]: You need to validate the index like in all other PopupMenu setters:
```
if (p_idx < 0) {
	p_idx += get_item_count();
}
ERR_FAIL_INDEX(p_idx, items.size());
```

The CI caught a crash here:
```
GDSCRIPT CODE:     PopupMenu.new().set_item_icon_modulate(100, Color(100, 100, 100, 1))
ERROR: FATAL: Index p_index = 100 is out of bounds (size() = 0).
   at: get (./core/templates/cowdata.h:155)
```

Comment (the-sink) [reply scene/gui/popup_menu.cpp:1260]: Oops, didn't notice during the rebase, thanks!

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**